awx/main/scheduler/task_manager.py

```diff
@@ -686,14 +687,27 @@ def task_needs_capacity(self, task, tasks_to_update_job_explanation):
                 tasks_to_update_job_explanation.append(task)
         logger.debug("{} couldn't be scheduled on graph, waiting for next cycle".format(task.log_format))
 
-    def reap_jobs_from_orphaned_instances(self):
-        # discover jobs that are in running state but aren't on an execution node
+    def reap(self):
+        # Finalize deprovisioned nodes that are no longer running jobs.
+        # Also, discover jobs that are in running state but aren't on an execution node
         # that we know about; this is a fairly rare event, but it can occur if you,
         # for example, SQL backup an awx install with running jobs and restore it
         # elsewhere
-        for j in UnifiedJob.objects.filter(
-            status__in=['pending', 'waiting', 'running'],
-        ).exclude(execution_node__in=Instance.objects.exclude(node_type='hop').values_list('hostname', flat=True)):
+
+        instances = Instance.objects.only('hostname', 'node_state')
+        jobs = UnifiedJob.objects.filter(status__in=['pending', 'waiting', 'running']).select_related('instance_group')
```

Comment:
This one line is my hangup. The last development cycle saw a massive improvement in the job lifecycle, and the time until the dispatcher starts a job. This code lines in that critical path, as any delay here is a delay submitting to pg_notify. Before this, the `.exclude` would at least prevent returning the bulk of the active jobs (essentially all of them if we're not in an error state). There are a lot of hazards.